share/man/man1/pegasus-cluster.1
'\" t
.\"     Title: pegasus-cluster
.\"    Author: [see the "Authors" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 11/25/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PEGASUS\-CLUSTER" "1" "11/25/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pegasus-cluster \- run a list of applications
.SH "SYNOPSIS"
.sp
.nf
\fBpegasus\-cluster\fR [\fB\-d\fR] [\fB\-e\fR | \fB\-f\fR] [\fB\-S ec\fR] [\fB\-s fn\fR] [\fB\-R fn\fR] [\fB\-n nr\fR] [\fBinputfile\fR]
.fi
.SH "DESCRIPTION"
.sp
The \fBpegasus\-cluster\fR tool executes a list of application in the order specified (assuming sequential mode\&.) It is generally used to do horizontal clustering of independent application, and does not care about any application failures\&. Such failures should be caught by using \fBpegasus\-kickstart\fR to start application\&.
.sp
In vertical clustering mode, the \fIhard failure\fR mode is encouraged, ending execution as soon as one application fails\&. When running a complex workflow through \fBpegasus\-cluster\fR , the order of applications in the input file must be topologically sorted\&.
.sp
Applications are usually using \fBpegasus\-kickstart\fR to execute\&. In the \fBpegasus\-kickstart\fR case, all invocations of \fBpegasus\-kickstart\fR except the first should add the \fBpegasus\-kickstart\fR option \fI\-H\fR to supress repeating the XML preamble and certain other headers of no interest when repeated\&.
.sp
\fBpegasus\-cluster\fR permits shell\-style quoting\&. One level of quoting is removed from the arguments\&. Please note that \fBpegasus\-kickstart\fR will also remove one level of quoting\&.
.SH "ARGUMENTS"
.PP
\fB\-d\fR
.RS 4
This option increases the debug level\&. Debug message are generated on
\fIstdout\fR
\&. By default, debugging is minimal\&.
.RE
.PP
\fB\-e\fR
.RS 4
This flag turns on the old behavior of
\fBpegasus\-cluster\fR
to always run everything
\fIand\fR
return success no matter what\&. The
\fB\-e\fR
flag is mutually exclusive with the
\fB\-f\fR
flag\&. By default, all applications are executed regardles of failures\&. Any detected application failure results in a non\-zero exit status from
\fBpegasus\-cluster\fR\&.
.RE
.PP
\fB\-f\fR
.RS 4
In hard failure mode, as soon as one application fails, either through a non\-zero exit code, or by dying on a signal, further execution is stopped\&. In parallel execution mode, one or more other applications later in the sequence file may have been started already by the time failure is detected\&.
\fBPegasus\-cluster\fR
will wait for the completion of these applications, but not start new ones\&. The
\fB\-f\fR
flag is mutually exclusive with the
\fB\-e\fR
flag\&. By default, all applications are executed regardless of failures\&. Any detected application failure results in a non\-zero exit status from
\fBpegasus\-cluster\fR\&.
.RE
.PP
\fB\-h\fR
.RS 4
This option prints the help message and exits the program\&.
.RE
.PP
\fB\-s fn\fR
.RS 4
This option will send protocol message (for Mei) to the specified file\&. By default, all message are written to
\fIstdout\fR
\&.
.RE
.PP
\fB\-R fn\fR
.RS 4
The progress reporting feature, if turned on, will write one event record whenever an application is started, and one event record whenever an application finished\&. This is to enable tracking of jobs in progress\&. By default, track logs are not written, unless the environment variable
\fISEQEXEC_PROGRESS_REPORT\fR
is set\&. If set, progress reports are appended to the file pointed to by the environment variable\&.
.RE
.PP
\fB\-S ec\fR
.RS 4
This option is a multi\-option, which may be used multiple times\&. For each given non\-zero exit\-code of an application, mark it as a form of success\&. In
\fB\-f\fR
mode, this means that
\fBpegasus\-cluster\fR
will not fail when seeing this exit code from any application it runs\&. By default, all non\-zero exit code constitute failure\&.
.RE
.PP
\fB\-n nr\fR
.RS 4
This option determines the amount of parallel execution\&. Typically, parallel execution is only recommended on multi\-core systems, and must be deployed rather carefully, i\&.e\&. only completely independent jobs across of whole
\fIinputfile\fR
should ever be attempted to be run in parallel\&. The argument
\fBnr\fR
is the number of parallel jobs that should be used\&. In addition to a non\-negative integer, the word
\fIauto\fR
is also understood\&. When
\fIauto\fR
is specified,
\fBpegasus\-cluster\fR
will attempt to automatically determine the number of cores available in the system\&. Strictly sequential execution, as if
\fInr\fR
was 1, is the default\&. If the environment variable
\fISEQEXEC_CPUS\fR
is set, it will determine the default number of CPUs\&.
.RE
.PP
\fBinputfile\fR
.RS 4
The input file specifies a list of application to run, one per line\&. Comments and empty lines are permitted\&. The comment character is the octothorpe (#), and extends to the end of line\&. By default,
\fBpegasus\-cluster\fR
uses
\fIstdin\fR
to read the list of applications to execute\&.
.RE
.SH "RETURN VALUE"
.sp
The \fBpegasus\-cluster\fR tool returns 1, if an illegal option was used\&. It returns 2, if the status file from option \fB\-s\fR cannot be opened\&. It returns 3, if the input file cannot be opened\&. It does \fInot\fR return any failure for failed applications in old\-exit \fB\-e\fR mode\&. In \fIdefault\fR and hard failure \fB\-f\fR mode, it will return 5 for true failure\&. The determination of failure is modified by the \fB\-S\fR option\&.
.sp
All other internal errors being absent, \fBpegasus\-cluster\fR will always return 0 when run without \fB\-f\fR \&. Unlike shell, it will \fInot\fR return the last application\(cqs exit code\&. In \fIdefault\fR mode, it will return 5, if any application failed\&. Unlike shell, it will \fInot\fR return the last application\(cqs exit code\&. However, it will execute all applications\&. The determination of failure is modified by the \fB\-S\fR flag\&. In \fB\-f\fR mode, *pegasus\-cluster returns either 0 if all main sequence applications succeeded, or 5 if one failed; or more than one in parallel execution mode\&. It will run only as long as applications were successful\&. As before, the *\-S flag determines what constitutes a failure\&.
.sp
The \fBpegasus\-cluster\fR application will also create a small summary on \fIstdout\fR for each job, and one for itself, about the success and failure\&. The field \fBfailed\fR reports any exit code that was not zero or a signal of death termination\&. It does \fInot\fR include non\-zero exit codes that were marked as success using the \fB\-S\fR option\&.
.SH "TASK SUMMARY"
.sp
Each task executed by \fBpegasus\-cluster\fR generates a record bracketed by square brackets like this (each entry is broken over two lines for readability):
.sp
.if n \{\
.RS 4
.\}
.nf
[cluster\-task id=1, start="2011\-04\-27T14:31:25\&.340\-07:00", duration=0\&.521,
 status=0, line=1, pid=18543, app="/bin/usleep"]
[cluster\-task id=2, start="2011\-04\-27T14:31:25\&.342\-07:00", duration=0\&.619,
 status=0, line=2, pid=18544, app="/bin/usleep"]
[cluster\-task id=3, start="2011\-04\-27T14:31:25\&.862\-07:00", duration=0\&.619,
 status=0, line=3, pid=18549, app="/bin/usleep"]
.fi
.if n \{\
.RE
.\}
.sp
Each record is introduced by the string \fIcluster\-task\fR with the following constituents, where strings are quoted:
.PP
\fBid\fR
.RS 4
This is a numerical value for main sequence application, indicating the application\(cqs place in the sequence file\&. The setup task uses the string
\fIsetup\fR
, and the cleanup task uses the string
\fIcleanup\fR
\&.
.RE
.PP
\fBstart\fR
.RS 4
is the ISO 8601 time stamp, with millisecond resolution, when the application was started\&. This string is quoted\&.
.RE
.PP
\fBduration\fR
.RS 4
is the application wall\-time duration in seconds, with millisecond resolution\&.
.RE
.PP
\fBstatus\fR
.RS 4
is the
\fIraw\fR
exit status as returned by the
\fIwait\fR
family of system calls\&. Typically, the exit code is found in the high byte, and the signal of death in the low byte\&. Typically, 0 indicates a successful execution, and any other value a problem\&. However, details could differ between systems, and exit codes are only meaningful on the same os and architecture\&.
.RE
.PP
\fBline\fR
.RS 4
is the line number where the task was found in the main sequence file\&. Setup\- and cleanup tasks don\(cqt have this attribute\&.
.RE
.PP
\fBpid\fR
.RS 4
is the process id under which the application had run\&.
.RE
.PP
\fBapp\fR
.RS 4
is the path to the application that was started\&. As with the progress record, any
\fBpegasus\-kickstart\fR
will be parsed out so that you see the true application\&.
.RE
.SH "PEGASUS-CLUSTER SUMMARY"
.sp
The final summary of counts is a record bracketed by square brackets like this (broken over two lines for readability):
.sp
.if n \{\
.RS 4
.\}
.nf
[cluster\-summary stat="ok", lines=3, tasks=3, succeeded=3, failed=0, extra=0,
 duration=1\&.143, start="2011\-04\-27T14:31:25\&.338\-07:00", pid=18542, app="\&./seqexec"]
.fi
.if n \{\
.RE
.\}
.sp
The record is introduced by the string \fIcluster\-summary\fR with the following constituents:
.PP
\fBstat\fR
.RS 4
The string
\fIfail\fR
when
\fBpegasus\-cluster\fR
would return with an exit status of 5\&. Concretely, this is any failure in
\fIdefault\fR
mode, and first failure in
\fB\-f\fR
mode\&. Otherwise, it will always be the string
\fIok\fR
, if the record is produced\&.
.RE
.PP
\fBlines\fR
.RS 4
is the stopping line number of the input sequence file, indicating how far processing got\&. Up to the number of cores additional lines may have been parsed in case of
\fB\-f\fR
mode\&.
.RE
.PP
\fBtasks\fR
.RS 4
is the number of tasks processed\&.
.RE
.PP
\fBsucceeded\fR
.RS 4
is the number of main sequence jobs that succeeded\&.
.RE
.PP
\fBfailed\fR
.RS 4
is the number of main sequence jobs that failed\&. The failure condition depends on the
\fB\-S\fR
settings, too\&.
.RE
.PP
\fBextra\fR
.RS 4
is 0, 1 or 2, depending on the existence of setup\- and cleanup jobs\&.
.RE
.PP
\fBduration\fR
.RS 4
is the duration in seconds, with millisecond resolution, how long *pegasus\-cluster ran\&.
.RE
.PP
\fBstart\fR
.RS 4
is the start time of
\fBpegasus\-cluster\fR
as ISO 8601 time stamp\&.
.RE
.SH "SEE ALSO"
.sp
\fBpegasus\-kickstart(1)\fR
.SH "CAVEATS"
.sp
The \fB\-S\fR option sets success codes globally\&. It is not possible to activate success codes only for one specific application, and doing so would break the shell compatibility\&. Due to the global nature, use success codes sparingly as last resort emergency handler\&. In better plannable environments, you should use an application wrapper instead\&.
.SH "EXAMPLE"
.sp
The following shows an example input file to \fBpegasus\-cluster\fR making use of \fBpegasus\-kickstart\fR to track applications\&.
.sp
.if n \{\
.RS 4
.\}
.nf
#
# mkdir
/path/to/pegasus\-kickstart \-R HPC \-n mkdir /bin/mkdir \-m 2755 \-p split\-corpus split\-ne\-corpus
#
# drop\-dian

RELEASE_NOTES
$Id$
===============================
Release Notes for PEGASUS 4.3.1
===============================
We are happy to announce the release of Pegasus 4.3.1 .  

Pegasus 4.3.1 is a minor release, that has minor enhancements and fixes bugs to Pegasus 4.3.0 release. 
 
Enhancements:
--------------

1) Support for Fixed Output Mapper

   Using this output mapper, users can specify  an externally
   accessible URL in the properties file, pointing to a directory
   where the output files needs to be transferred to. To use this
   mapper, set the following  properties  
 
	pegasus.dir.storage.mapper Fixed
	pegasus.dir.storage.mapper.fixed.url  <url to the storage directory e.g. gsiftp://outputs.isi.edu/shared/outputs>
 

Bugs Fixed:
-------------- 

1) pegasus-analyzer does not detect jobs that are condor_rm'ed if no
postscript is associated with the job 

   By default, each job has a postscript associated that detects an
   empty job stdout and flags it as a failure. However, if a job is
   not asscociated with a postscript and a user/system condor_rm's the
   job, the failure is not detected. This is now fixed, and a
   JOB_ABORTED event is logged in the stampede database, when a job is
   aborted. 
 
2) IRODS support in pegasus-transfer

   The IRODS support in pegasus-transfer was broken. This is now fixed.
 
3) pegasus-kickstart compilation warnings for character encodings

   kickstart maintains a table to escape characters correctly for
   putting them in a XML document. The non-ascii characters in the
   table were latin1 not UTF-8. This caused a warning on newer
   versions of gcc, which could not be disabled across all platforms.
   kickstart now writes out it;s output in UTF-8 encoding and the xml
   escaping was updated accordingly. 
 
4) Fix to URL handling in the planner

   Changed the regex in PegasusURL so that we can pass urls with just
   the hostnames specified. e.g http://isis.isi.edu Note: no trailing
   / . Before the fix, the planner was throwing an exception if a user
   specified an input URL with path names containing only one
   directory. 
 
5) planner had a rogue debug statement

   There was a rogue system.out statement in the planner output that
   led to a statement being logged for each job in the workflow. 
 
6) pegasus-statistics had 2.6 style code

   Pegasus is distributed as part of OSG software stack, and one of
   the supported platforms there is EL5 systems that come with Python
   2.4. pegasus-statistics had some code that was compatible with
   2.5. This is now fixed.

===============================
Release Notes for PEGASUS 4.3
===============================
We are happy to announce the release of Pegasus 4.3.  

Pegasus 4.3 is a major release of Pegasus which contains all the
enhancements and bugfixes in 4.2.2 release 

New features and Improvements in 4.3 include

    - improvements to pegasus lite and optimizations for input file
    staging in non shared filesystem deployments 
    - support for output mappers, allowing users finer grained control
    over where to place the outputs on an output site 
    - support for SSH based submissions on top of Condor BOSCO. 
    - substantial improvements to pegasus-kickstart including abiltiy
    to track peak memory usage for jobs 
    - improvements to pegasus-s3 and pegasus-transfer


NEW FEATURES
--------------

1) Support for bypassing transfer of input files via the staging site 

   In the non shared filesystem deployments (
   pegasus.data.configuration = nonsharedfs|condorio) users, can now
   setup pegasus to transfer the input files directly to the worker
   nodes without going through the staging site.  This can be done by
   setting the following property to true 
  
   pegasus.transfer.bypass.input.staging

   In the nonsharedfs case, if the input files are already present on
   a shared disk accessible from the worker nodes, pegasus lite can
   symlink instead of copying them over to the local directory on the
   worker node.  The cleanup algorithm was updated to ignore files
   that are directly pulled to the worker nodes from the input site
   locations. 
 
2) Support for DAX generation jobs in hierarchal workflows

   Pegasus now has support for having a dax generation job in the
   workflow. This allows users to add long running dax generation jobs
   as a compute job in the workflow, that can be run remotely. These
   dax generation jobs need to be a parent of the associated DAX
   job. Pegasus will ensure that the DAX generated on a remote site is
   brought to the local site for the associated sub workflow
   corresponding to the DAX job to be planned. 

   Earlier, the only way for hierarchal workflows was that the DAX'es
   for the sub workflows had to be pre generated and the paths to the
   dax files was specified in the DAX jobs.  Pegasus did not
   automatically handle separate DAX generations jobs out of the
   box. More details can be found at
   https://jira.isi.edu/browse/PM-667. There is an example checked in
   the share/pegasus/examples/dynamic-hierarchy directory. 
 
3) Support for output mappers
   
   Pegasus now has support for output mappers, that allow users fine
   grained control over how the output files on the output site are
   laid out.  The mappers can be configured by setting the following
   property  

    pegasus.dir.storage.mapper

   The following mappers are supported

   Flat: By default, Pegasus will place the output files in the
   storage directory specified in the site catalog for the output
   site. 

   Hashed: This mapper results in the creation of a deep directory
   structure on the output site, while populating the results. The
   base directory on the remote end is determined from the site
   catalog. Depending on the number of files being staged to the
   remote site a Hashed File Structure is created that ensures that

src/tools/pegasus-cluster/pegasus-cluster.c
/*
 * This file or a portion of this file is licensed under the terms of
 * the Globus Toolkit Public License, found in file GTPL, or at
 * http://www.globus.org/toolkit/download/license.html. This notice must
 * appear in redistributions of this file, with or without modification.
 *
 * Redistributions of this Software, with or without modification, must
 * reproduce the GTPL in: (1) the Software, or (2) the Documentation or
 * some other similar material which is provided with the Software (if
 * any).
 *
 * Copyright 1999-2004 University of Chicago and The University of
 * Southern California. All rights reserved.
 *
 */
#include <sys/types.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <sys/time.h>
#include <sys/resource.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <unistd.h>

#include "tools.h"
#include "parser.h"
#include "report.h"
#include "mysystem.h"
#include "job.h"
#include "statinfo.h"

#ifndef APPLICATION_NAME
#define APPLICATION_NAME "pegasus-cluster"
#endif /* APPLICATION_NAME */

#ifndef RECORD_PREFIX
#define RECORD_PREFIX "cluster"
#endif /* RECORD_PREFIX */


extern char *optarg;
extern int optind, opterr, optopt;
int debug = 0;
int progress = -1;
char* application = APPLICATION_NAME;
static char success[257];

static
void
helpMe( const char* programname, int rc )
/* purpose: write help message and exit
 * paramtr: programname (IN): application of the program (us)
 *           rc (IN): exit code to exit with
 * returns: procedure does not return. */
{
  printf( "Usage:\t%s [options] [inputfile]\n\n", programname );
  printf( "Optional arguments:\n"
" -d\tIncrease debug mode.\n"
" -s fn\tProtocol anything to given status file, default stdout.\n"
" -R fn\tRecords progress into the given file, see also SEQEXEC_PROGRESS_REPORT.\n"
" -S ec\tMulti-option: Mark non-zero exit-code ec as success.\n"
" -n nr\tNumber of CPUs to use, defaults to 1, string 'auto' permitted.\n"
" input\tFile with list of applications and args to execute, default stdin.\n\n"
"Execution control and exit code:\n"
"\tExecute everything but return success only if all were successful.\n"
" -e\tExecute everything (old default mode) and always return success.\n"
" -f\tFail hard on first error (non-zero exit code or signal death).\n"
"\tOption -e and -f are mutually exclusive.\n" );
  exit(rc);
}

static
int
processors( void )
{
  long config = 
#ifdef _SC_NPROCESSORS_CONF
    sysconf( _SC_NPROCESSORS_CONF )
#else
    1
#endif 
    ; 

  long online = 
#ifdef _SC_NPROCESSORS_ONLN
    sysconf( _SC_NPROCESSORS_ONLN )
#else
    1 
#endif
    ;

  if ( config <= 0 ) config = 1; 
  if ( online <= 0 ) online = 1; 
  return config < online ? config : online;
}

static
void
parseCommandline( int argc, char* argv[], 
		  int* fail_hard, int* old_mode, int* cpus )
{
  char *s, *ptr = strrchr( argv[0], '/' );
  int option, tmp;

  /* exit code 0 is always good, just in case */
  memset( success, 0, sizeof(success) );
  success[0] = 1;
  *cpus = 1; 

  if ( ptr == NULL ) ptr = argv[0];
  else ptr++;
  application = ptr;

  /* default progress report location from environment variable */
  if ( (s = getenv("SEQEXEC_PROGRESS_REPORT")) != NULL ) {
    if ( (progress = open( s, O_WRONLY | O_APPEND | O_CREAT, 0666 )) == -1 ) {
      showerr( "%s: open progress %s: %d: %s\n",
	       application, s, errno, strerror(errno) );
    }
  }

  /* default parallelism from environment variable */ 
  if ( (s = getenv("SEQEXEC_CPUS")) != NULL ) { 
    if ( strcasecmp( s, "auto" ) == 0 ) *cpus = processors();
    else *cpus = atoi(s); 
    if ( *cpus < 0 ) *cpus = 1; 
  }

  opterr = 0;
  while ( (option = getopt( argc, argv, "R:S:defhn:s:" )) != -1 ) {
    switch ( option ) {
    case 'R':
      if ( progress != -1 ) close(progress);
      if ( (progress = open( optarg, O_WRONLY | O_APPEND | O_CREAT, 0666 )) == -1 ) {
	showerr( "%s: open progress %s: %d: %s\n",
		 application, optarg, errno, strerror(errno) );
      }
      break;

    case 'S':
      tmp = atoi(optarg);
      if ( tmp > 0 && tmp < sizeof(success) ) success[tmp] = 1;
      else showerr( "%s: Ignoring unreasonable success code %d\n", application, tmp );
      break;

    case 'd':
      debug++;
      break;

    case 'e':
      *old_mode = 1;
      *fail_hard = 0;
      break;

    case 'f':
      *fail_hard = 1;
      *old_mode = 0;
      break;

    case 'n':
      if ( strcasecmp( optarg, "auto" ) == 0 ) *cpus = processors();
      else *cpus = atoi(optarg); 
      if ( *cpus < 0 ) *cpus = 1; 
      break;

    case 's':
      if ( freopen( optarg, "w", stdout ) == NULL ) {
	showerr( "%s: open status %s: %d: %s\n",
		 application, optarg, errno, strerror(errno) );
	exit(2);
      }
      break;

    case 'h':
    case '?':
      helpMe( ptr, 0 );
      break;

    default:
      helpMe( ptr, 1 );
      break;
    }
  }

  if ( (argc - optind) > 1 ) helpMe( ptr, 1 );
  if ( argc != optind ) {
    if ( (freopen( argv[optind], "r", stdin )) == NULL ) {
      showerr( "%s: open input %s: %d: %s\n",
	       application, argv[optind], errno, strerror(errno) );
      exit(3);
    }
  }
}

static
char*
merge( char* s1, char* s2, int use_space )
/* purpose: merge two strings and return the result
 * paramtr: s1 (IN): first string, may be NULL
 *          s2 (IN): second string, must not be null
 *          use_space (IN): flag, if true, separate by space
 * returns: merge of strings into newly allocated area.
 */
{
  if ( s1 == NULL ) {
    return strdup(s2);
  } else {
    size_t l1 = strlen(s1);
    size_t l2 = strlen(s2);
    size_t len = l1 + l2 + 2;
    char* temp = (char*) malloc(len);
    strncpy( temp, s1, len );
    if ( use_space ) strncat( temp, " ", len );
    strncat( temp, s2, len );

    return temp;
  }
}

pid_t
wait_for_child( Jobs* jobs, int* status )
{
  struct rusage usage; 
  Signals save;
  int saverr; 
  double final; 
  size_t slot; 
  pid_t child = ((pid_t) -1);  

  /* FIXME: Not sure, if I need these. I take it to mean: While I am
   * blocked in kernel wait()ing for children, do not interrupt ^C me,
   * and do not send me SIGCHLD since I am inside wait() anyways. Do
   * send the signals to the children, though (which hopefully exit.)
   */ 
  save_signals(&save); 
  errno = 0; /* we rely later on wait4 results */ 
  while ( (child = wait4( ((pid_t) 0), status, 0, &usage )) < 0 ) {
    saverr = errno;
    perror( "wait4" ); 
    errno = saverr; 

    if ( errno != EINTR ) {
      *status = -1;
      break;
    }
  }
  saverr = errno; 
  final = now(NULL); 

  /* FIXME: see above, end bracket. */ 
  restore_signals(&save); 

  /* find child that has finished */ 
  for ( slot=0; slot < jobs->cpus; ++slot ) {
    if ( jobs->jobs[slot].child == child ) break;
  }

  if ( slot == jobs->cpus ) { 
    /* reaped child not found, not good */
    showerr( "%s: process %d (status %d) is not a known child, ignoring.\n", 
	     application, child, *status ); 
  } else { 
    /* free slot and report */ 
    char date[32]; 
    Job* j = (jobs->jobs) + slot; 

    /* 20110419 PM-364: new requirement */
    showout( "[%s-task id=%lu, start=\"%s\", duration=%.3f, status=%d, "
	     "line=%lu, pid=%d, app=\"%s\"]\n", RECORD_PREFIX,
	     j->count,
	     iso2date( j->start, date, sizeof(date) ),
	     (final - j->start), 
	     *status,
	     j->lineno, 
	     child, 
	     j->argv[ find_application(j->argv) ] ); 
    
    /* progress report at finish of job */ 
    if ( progress != -1 ) 
      report( progress, j->start, (final - j->start), *status, j->argv, &usage, NULL 
#ifndef MONOTONICALLY_INCREASING
	    , j->count
#endif /* MONOTONICALLY_INCREASING */
	    ); 
    
    /* free reported job */ 
    job_done(j); 
  }	   

  errno = saverr; 
  return child; 
}

void
run_independent_task( char* cmd, char* envp[], unsigned long* extra,
		      const char* special ) 
{
  if ( cmd != NULL ) { 
#ifndef USE_SYSTEM_SYSTEM
    size_t len; 
    int    appc, other;
    char** appv;

    if ( (appc = interpreteArguments( cmd, &appv )) > 0 ) {
      /* determine full path to application according to PATH */ 
      char* fqpn = find_executable( appv[0] ); 
      if ( fqpn ) {
	/* found a FQPN, exchange first item in argument vector */ 
	free((void*) appv[0]); 
	appv[0] = fqpn; 
      }

      other = mysystem( appv, envp, special ); 
      if ( other || debug )
	showerr( "%s: %s returned %d/%d\n", application, special,
		 (other >> 8), (other & 127) ); 
      for ( len=0; len<appc; len++ ) free((void*) appv[len]);
      free((void*) appv); 
    } else {
      /* unparsable cleanup argument string */
      showerr( "%s: unparsable %s string, ignoring\n", 
	       application, special ); 
    }
#else
    int other = system( cmd ); 
    if ( other || debug )
      showerr( "%s: %s returned %d/%d\n", application, special,
	       (other >> 8), (other & 127) ); 
#endif /* USE_SYSTEM_SYSTEM */
    (*extra)++; 
  }
}

int
isafailure( int status )
{
  /* FIXME: On systems with exit codes outside 0..256 this may core dump! */
  return ( WIFEXITED(status) && success[ WEXITSTATUS(status) ] == 1 ) ? 0 : 1;
}

void
massage_failure( int fail_hard, int current_ec, int* collect_ec )
{
  if ( fail_hard ) { 
    /* only propagate first failure in hard-fail mode */ 
    if ( ! ( *collect_ec && isafailure(*collect_ec) ) ) *collect_ec = current_ec; 
  } else {
    /* always retain last exit code in no-hard-fail mode */ 
    *collect_ec = current_ec; 
  }
}

int
main( int argc, char* argv[], char* envp[] )
{
  size_t len;
  char line[MAXSTR];
  int other, exitstatus, status = 0;
  int slot, cpus, fail_hard = 0, old_mode = 0; 
  char* cmd;
  char* save = NULL;
  unsigned long total = 0;
  unsigned long failure = 0;
  unsigned long lineno = 0;
  unsigned long extra = 0; 
  time_t when;
  Jobs jobs;
  double diff, start = now(&when);
  parseCommandline( argc, argv, &fail_hard, &old_mode, &cpus );
  
  /* progress report finish */
  if ( progress != -1 ) {
    report( progress, start, 0.0, -1, argv, NULL, NULL
#ifndef MONOTONICALLY_INCREASING
	  , 0ul
#endif /* MONOTONICALLY_INCREASING */
	  );
  }

  /* allocate job management memory */ 
  if ( jobs_init( &jobs, cpus ) == -1 ) {
    showerr( "%s: out of memory: %d: %s\n", 
	     application, errno, strerror(errno) );
    return 42;
  }
  
  /* since we will create multiple concurrent processes, let's create a
   * process group to order them by.
   */
  if ( setpgid( 0, 0 ) == -1 ) {
    showerr( "%s: unable to become process group leader: %d: %s (ignoring)\n", 
	     application, errno, strerror(errno) ); 
  }

  /* NEW: unconditionally run a setup job */
  run_independent_task( getenv("SEQEXEC_SETUP"), envp, &extra, "setup" ); 

  /* Read the commands and call each sequentially */
  while ( fgets(line,sizeof(line),stdin) != (char*) NULL ) {
    ++lineno;

    /* check for skippable line */
    if ( line[0] == 0 || /* empty line */
	 line[0] == '\r' || /* CR */
	 line[0] == '\n' || /* LF */
	 line[0] == '#' /* comment */ ) continue;

    /* check for unterminated line (line larger than buffer) */
    len = strlen(line);
    if ( line[len-1] != '\r' && line[len-1] != '\n' ) {
      /* read buffer was too small, save and append */
      char* temp = merge( save, line, 0 );
      if ( save != NULL ) free((void*) save);
      save = temp;

      lineno--;
      showerr( "%s: continuation line %lu\n", application, lineno ); 
      continue;
    } else {
      /* remove line termination character(s) */
      do { 
	line[len-1] = 0;
	len--;
      } while ( len > 0 && (line[len-1] == '\r' || line[len-1] == '\n') );
    }
	
    /* Assemble command.
     * FIXME: barf if commandline becomes too long, see _SC_ARG_MAX. 
     */
    if ( save != NULL ) {
      cmd = merge( save, line, 0 );
      free((void*) save);
      save = NULL;
    } else {
      cmd = line;
    }

    /* find a free slot */ 
    while ( (slot = jobs_first_slot( &jobs, EMPTY )) == jobs.cpus ) {
      /* wait for any child to finish */
      if ( debug ) showerr( "%s: %d slot%s busy, wait()ing\n", 
			    application, jobs.cpus, ( jobs.cpus == 1 ? "" : "s" ) ); 
      wait_for_child( &jobs, &other ); 
      if ( errno == 0 && isafailure(other) ) failure++; 
      massage_failure( fail_hard, other, &status );
    }
    /* post-condition: there is a free slot; slot number in "slot" */ 

    /* found free slot */ 
    if ( fail_hard && status && isafailure(status) ) {
      /* we are in failure mode already, skip starting new stuff */ 
    } else if ( slot < jobs.cpus ) { 
      /* there is a free slot. Spawn and continue */ 
      Signals save; 
      Job* j = jobs.jobs + slot; 
      if ( (j->argc = interpreteArguments( cmd, &(j->argv) )) > 0 ) {
	/* determine full path to application according to PATH */ 
	char* fqpn = find_executable( j->argv[0] ); 
	if ( fqpn ) { 
	  /* found a FQPN, exchange first item in argument vector */ 
	  free((void*) j->argv[0]); 
	  j->argv[0] = fqpn; 
	}

	total++;
	j->envp = envp; /* for now */ 
	j->lineno = lineno; 

	/* WARNING: Must propagate "save" to start_child() */
	save_signals( &save ); 

	if ( (j->child = fork()) == ((pid_t) -1) ) { 
	  /* fork error, bad */
	  showerr( "%s: fork: %d: %s\n", 
		   application, errno, strerror(errno) ); 
	  failure++; 
	  job_done( j ); 
	} else if ( j->child == ((pid_t) 0) ) {
	  /* child code */
	  start_child( j->argv, j->envp, &save ); 
	  return 127; /* never reached, just in case */ 
	} else {
	  /* parent code */
	  j->count = total; 
	  j->state = RUNNING;
	  j->start = now( &(j->when) ); 
	}

	/* END BRACKET */ 
	restore_signals( &save ); 

      } else {
	/* error parsing args */
	if ( debug ) 
	  showerr( "%s: error parsing arguments on line %lu, ignoring\n", 
		   application, lineno ); 
      }
    } else {
      /* no free slots, wait for children to finish */ 
      showerr( "%s: %s:%d THIS SHOULD NOT HAPPEN! (ignoring)\n", 
	       application, __FILE__, __LINE__ ); 
    }

    if ( cmd != line ) free((void*) cmd);

    /* fail hard mode, if requested */
    if ( fail_hard && status && isafailure(status) ) break;
  }

  /* wait for all children */ 
  while ( (slot=jobs_in_state( &jobs, EMPTY )) < jobs.cpus ) {
    /* wait for any child to finish */
    size_t n = jobs.cpus - slot; 
    if ( debug ) showerr( "%s: %d task%s remaining\n", application, 
			  n, (n == 1 ? "" : "s" ) ); 
    wait_for_child( &jobs, &other );
    if ( errno == 0 && isafailure(other) ) failure++; 
    massage_failure( fail_hard, other, &status );
  }

  /* NEW: unconditionally run a clean-up job */
  run_independent_task( getenv("SEQEXEC_CLEANUP"), envp, &extra, "cleanup" ); 

  /* compute if seqexec should return any form of failure */
  if ( old_mode ) { 
    exitstatus = 0; 
  } else if ( fail_hard ) {
    exitstatus = (status && isafailure(status)) ? 5 : 0;
  } else {
    exitstatus = failure ? 5 : 0;
  }

  /* provide final statistics */
  jobs_done( &jobs ); 
  diff = now(NULL) - start;
  showout( "[%s-summary stat=\"%s\", lines=%lu, tasks=%lu, succeeded=%lu, failed=%lu, "
	   "extra=%lu, duration=%.3f, start=\"%s\", pid=%d, app=\"%s\"]\n",
	   RECORD_PREFIX,
	   exitstatus ? "fail" : "ok",
	   lineno, total, total-failure, failure, extra,
	   diff, iso2date(start,line,sizeof(line)),
	   getpid(), argv[0] );

  fflush(stdout); /* just in case */
  exit( exitstatus ); 
}
